Rewrite and restructure vector reference docs (#434)
redis/docs: content/develop/interact/search-and-query/advanced-concepts/vectors.md

| `vector_blob_param`  | The query vector as bytes and must be passed through the `PARAMS` section. The blob's byte size should match the vector field dimension and type.  |
| `vector_query_params` (optional) | An optional section for marking one or more vector query parameters passed through the `PARAMS` section. Valid parameters should be provided as key-value pairs. See which [runtime query params]({{< baseurl >}}/develop/interact/search-and-query/advanced-concepts/vectors#runtime-query-params) are supported for each vector index type.  |
| `distance_field` (optional) | The optional distance field name used in the response and/or for sorting. By default, the distance field name is "`__<vector_field>_score`" and it can be used for sorting without using `AS <distance_field>` in the query.  |
| `vector_query_params_count` | The number of vector query params.  |
| `vector_query_param_name` | The name of the vector query param.  |
| `vector_query_param_value` | The value of the vector query param.  |

**Example**

```
FT.SEARCH my_index "(@title:Matrix @year:[2020 2022])=>[KNN 10 @vector $BLOB]" PARAMS 2 BLOB "\x12\xa9\xf5\x6c" DIALECT 2
```

**Use query attributes**

> Alternatively, as of v2.6, `<vector_query_params>` and `<distance_field>` name can be specified in runtime

Author: I want product @joeywhelan et al to comment on this because I believe we are going to be shifting how we discuss versions. This is the nomenclature present in the docs before the overhaul

[query attributes]({{< relref "/develop/interact/search-and-query/advanced-concepts/query_syntax" >}}#query-attributes) as well. Thus, the following format is also supported:

```
[KNN <top_k> @<vector_field> $<vector_blob_param>]=>{$yield_distance_as: <distance_field>}
```

### Vector range queries
Vector range queries allow you to filter the index using a `radius` parameter representing the semantic distance between an input query vector and indexed vector fields. This is useful in scenarios when you don't know the right value to use for `top_k`, but you know what distance threshold should be applied.

Vector range queries operate slightly different than standard vector queries:

> - Vector range queries can appear *multiple times* in a query as a filter criteria.

Author: Will add some additional context.

| `index_name`  | Name of the index.  |
| `vector_field`  | Name of the vector field in the index. |
| `radius` or `radius_param` | The maximum semantic distance allowed between the query vector and indexed vectors. You can provide the value directly in the query, passed to the `PARAMS` section, or as a query attribute.
| `vector_blob_param`  | The query vector as bytes and must be passed through the `PARAMS` section. The blob's byte size should match the vector field dimension and type. |
| `vector_query_params` (optional) | An optional section for marking one or more vector query parameters passed through the `PARAMS` section. Valid parameters should be provided as key-value pairs. See which [runtime query params]({{< baseurl >}}/develop/interact/search-and-query/advanced-concepts/vectors#runtime-query-params) are supported for each vector index type.  |
| `vector_query_params_count` | The number of vector query params.  |
| `vector_query_param_name` | The name of the vector query param.  |
| `vector_query_param_value` | The value of the vector query param.  |

**Use query attributes**

> Vector range queries clause can be followed by a query attributes section as follows: `@<vector_field>: [VECTOR_RANGE (<radius> | $<radius_param>) $<vector_blob_param>]=>{$<param>: (<value> | $<value_attribute>); ... }`, where the relevant params in that case are `$yield_distance_as` and `$epsilon`. Note that there is no default distance field name in range queries.

Author: Yes

### Filters

> Redis supports vector search along with various filters to narrow the search space based on defined criteria. If your index contains searchable fields (for example, `TEXT`, `TAG`, `NUMERIC`, `GEO`, `GEOSHAPE`, and `VECTOR`), you can perform vector searches with filters.

Reviewer: This seems to imply that other VECTOR fields can act as filters. Is that true? What's a real-world use case for filtering on a vector field before performing a vector search? Does that actually narrow the search space effectively?

> Redis supports vector search along with various filters to narrow the search space based on defined criteria. If your index contains searchable fields (for example, `TEXT`, `TAG`, `NUMERIC`, `GEO`, `GEOSHAPE`, and `VECTOR`), you can perform vector searches with filters.

Author: Yes. Will add

> Redis supports vector search along with various filters to narrow the search space based on defined criteria. If your index contains searchable fields (for example, `TEXT`, `TAG`, `NUMERIC`, `GEO`, `GEOSHAPE`, and `VECTOR`), you can perform vector searches with filters.

Author: I am not sure we can claim that filters reduce query latency. @joeywhelan

**Supported filter types**

- [Exact Match Filters](https://redis.io/docs/develop/interact/search-and-query/query/exact-match/)
- [Numeric Range Filters](https://redis.io/docs/develop/interact/search-and-query/query/range/)
- [Full-text Search](https://redis.io/docs/develop/interact/search-and-query/query/full-text/)
- [Geospatial](https://redis.io/docs/develop/interact/search-and-query/query/geo-spatial/)

You can also [combine multiple queries](https://redis.io/docs/develop/interact/search-and-query/query/combined/) as a filter.

**Syntax**

Vector search queries with filters follow this basic structure:

```
FT.SEARCH <index_name> <primary_filter_query>=>[...]
```

where `<primary_filter_query>` defines document selection and filtering.

### How Filtering Works

> Redis uses internal algorithms to optimize the filtering computation for vector search. The runtime algorithm is determined by heuristics that aim to minimize query latency, based on several factors derived from the query and the index.

Reviewer: ```suggestion
Redis uses internal algorithms to optimize the filtering computation for vector search. The runtime algorithm is determined by heuristics that aim to minimize query latency based on several factors derived from the query and the index.
```

**Batches mode**

Batches mode works by paginating through small batches of nearest neighbors from the index:
- A batch of high-scoring documents from the vector index is retrieved. These documents are yielded ONLY if the `<primary_filter_query>` is satisfied. In other words, the document must contain a similar vector and meet the filter criteria.
- The iterative procedure terminates when `<top_k>` documents that pass the filter criteria are yielded, or after every vector in the index has been processed.
- The batch size is determined automatically by heuristics based on `<top_k>`, and the ratio between the expected number of documents in the index that pass the `<primary_filter_query>` and the vector index size.

> - The goal is to minimize the total number of batches required to get the `<top_k>` results while preserving a small batch size as possible. Note that the batch size may change dynamically in each iteration based on the number of results that passed the filter in previous batches.

Reviewer: ```suggestion
- The goal is to minimize the total number of batches required to get the `<top_k>` results while preserving the smalles batch size possible. Note that the batch size may change dynamically in each iteration based on the number of results that pass the filter in previous batches.
```

### Index-specific query parameters

**FLAT**

Currently, there are no runtime params available for FLAT indexes.

**HNSW**

Optional runtime params for HNSW indexes are:

| Parameter       | Description                                                                                               | Default value       |
|:----------------|:----------------------------------------------------------------------------------------------------------|:--------------------|
| `EF_RUNTIME`    | The maximum number of top candidates to hold during the KNN search. Higher values lead to more accurate results at the expense of a longer query runtime. | Value passed during index creation (default is 10). |
| `EPSILON`       | Relative factor that sets the boundaries for a vector range query. Vector candidates whose distance from the query vector is `radius * (1 + EPSILON)` are potentially scanned, allowing a more extensive search and more accurate results at the expense of runtime. | Value passed during index creation (default is 0.01). |

### Important notes

{{% alert title="Important notes" color="info" %}}

> 1. Although specifying `<top_k>` requested results in KNN search, the default `LIMIT` is 10, to get all the returned results, specify `LIMIT 0 <top_k>` in your search command. See examples below.

Reviewer: I'm having a really hard time parsing this sentence. I suggest reading it aloud. This should probably be broken up into a 2-3 shorter sentences, with clearer subjects and predicates.

Author: I think this is just a typo with an unintentional comma where there was supposed to be a period.

2. By default, the results are sorted by their document's score. To sort by some vector similarity score, use `SORTBY <distance_field>`. See examples below.

> 3. It is recommended to adjust the `<radius>` parameter in range queries to the corresponding vector field distance metric and to the data itself. In particular, recall that the distance between the vectors in an index whose distance metric is Cosine is bounded by `2`, while L2 distance between the vectors is not bounded. Hence, it is better to consider the distance between the vectors that are considered similar and choose `<radius>` accordingly.

Reviewer: Same comment for these setences. There are several grammatical issues. Also sudden shift in tone with phrases like "it is recommend that" and "hence". Bottom line: I don't really understand what's being communicated here.

{{% /alert %}}

## Vector search examples

> Below are a number of examples to help you get started. If you want a more comprehensive walkthrough, visit the vector [quickstart guide]({{< baseurl >}}/develop/get-started/vector-database) and the [Redis AI Resources](https://github.com/redis-developer/redis-ai-resources) repo for more assistance.

Reviewer: ```suggestion
Below are a number of examples to help you get started. If you want a more comprehensive walkthrough, see the [Redis vector quickstart guide]({{< baseurl >}}/develop/get-started/vector-database) and the [Redis AI Resources](https://github.com/redis-developer/redis-ai-resources) Github repo.
```

> ### Standard vector search examples

Reviewer: Again, what does "standard" mean here?

> Return the 10 nearest neighbor documents for which the vector stored under its `vector` field is the closest to the vector represented by the following 4-bytes blob:

Reviewer: Again, let's not call the vector field "vector". Makes the sentence very confusing to read.

Reviewer: Better: The query below returns the 10 documents whose "message_embedding" vector fields most closely match the provided 4-byte query vector. Here, the query vector is the byte string "\x12\xa9\xf5\x6c".

Of course, this requires renaming the vector field in the example.

Author: Point heard, making the update.

```
FT.SEARCH idx "*=>[KNN 10 @vector $BLOB]" PARAMS 2 BLOB "\x12\xa9\xf5\x6c" DIALECT 2
```
Now, sort the results by their distance from the query vector:
```
FT.SEARCH idx "*=>[KNN 10 @vector $BLOB]" PARAMS 2 BLOB "\x12\xa9\xf5\x6c" SORTBY __vector_score DIALECT 2
```

> Return the top 10 similar documents, use *query params* (see "params" section in [FT.SEARCH command]({{< baseurl >}}/commands/ft.search)) for specifying `K` and `EF_RUNTIME` parameter, and set `EF_RUNTIME` value to 150 (assuming `vector` is an HNSW index):

Reviewer: This needs to describe WHY the user would want to set these parameters.

Author: I will do a pass an update the language on the query examples. I had not really touched the descriptions yet.

```
FT.SEARCH idx "*=>[KNN $K @vector $BLOB EF_RUNTIME $EF]" PARAMS 6 BLOB "\x12\xa9\xf5\x6c" K 10 EF 150 DIALECT 2
```

> Similar to the previous queries, this time use a custom distance field name to sort by it:

Reviewer: ```suggestion
Now use use a custom distance field name to sort by it:
```

Reviewer: What's the use case? Not obvious.

```
FT.SEARCH idx "*=>[KNN $K @vector $BLOB AS my_scores]" PARAMS 4 BLOB "\x12\xa9\xf5\x6c" K 10 SORTBY my_scores DIALECT 2
```

> Use [query attributes]({{< baseurl >}}/develop/interact/search-and-query/advanced-concepts/query_syntax#query-attributes) syntax to specify optional parameters and the distance field name:

Reviewer: Why? What's the use case?

Author: Sophisticated query parametrization? I am not sure. It's a generic Redis query syntax construct in the API.  It was part of the previous docs and it's supported

```
FT.SEARCH idx "*=>[KNN 10 @vector $BLOB]=>{$EF_RUNTIME: $EF; $YIELD_DISTANCE_AS: my_scores}" PARAMS 4 EF 150 BLOB "\x12\xa9\xf5\x6c" SORTBY my_scores DIALECT 2
```

> Find additional Python examples for [`redis-py`](https://github.com/redis-developer/redis-ai-resources/blob/main/python-recipes/vector-search/00_redispy.ipynb) and [`redisvl`](https://github.com/redis-developer/redis-ai-resources/blob/main/python-recipes/vector-search/01_redisvl.ipynb).

Reviewer: ```suggestion
To see additional vector search examples, review the [`Redis Python client vector recipes`](https://github.com/redis-developer/redis-ai-resources/blob/main/python-recipes/vector-search/00_redispy.ipynb) and [`Redis Vector Library recipes`](https://github.com/redis-developer/redis-ai-resources/blob/main/python-recipes/vector-search/01_redisvl.ipynb) notebooks.
```

```
FT.SEARCH idx "(@type:{shirt} ~@color:{blue})=>[KNN $K @vector $BLOB]" PARAMS 4 BLOB "\x12\xa9\xf5\x6c" K 10 DIALECT 2
```

This example shows a pre-filter with KNN query in which the hybrid policy is set explicitly to "ad-hoc brute force" (rather than auto-selected):

```
FT.SEARCH idx "(@type:{shirt})=>[KNN $K @vector $BLOB HYBRID_POLICY ADHOC_BF]" PARAMS 4 BLOB "\x12\xa9\xf5\x6c" K 10 SORTBY __vec_scores DIALECT 2
```

This example shows a pre-filter with KNN query in which the hybrid policy is set explicitly to "batches", and the batch size is set explicitly to be 50 using a query parameter:

```
FT.SEARCH idx "(@type:{shirt})=>[KNN $K @vector $BLOB HYBRID_POLICY BATCHES BATCH_SIZE $B_SIZE]" PARAMS 6 BLOB "\x12\xa9\xf5\x6c" K 10 B_SIZE 50 DIALECT 2
```

Run the same query as above and use the query attributes syntax to specify optional parameters:

```
FT.SEARCH idx "(@type:{shirt})=>[KNN 10 @vector $BLOB]=>{$HYBRID_POLICY: BATCHES; $BATCH_SIZE: 50}" PARAMS 2 BLOB "\x12\xa9\xf5\x6c" DIALECT 2
```

> Find additional Python examples for [`redis-py`](https://github.com/redis-developer/redis-ai-resources/blob/main/python-recipes/vector-search/00_redispy.ipynb) and [`redisvl`](https://github.com/redis-developer/redis-ai-resources/blob/main/python-recipes/vector-search/01_redisvl.ipynb).

Reviewer: Is this text supposed to be same as the callout above?

- oss
- kubernetes
- clients
description: Learn how to use vector fields and perform vector searches in Redis
linkTitle: Vectors
math: true
title: Vectors
weight: 14
---

Redis is commonly used as a [highly-performant vector database](https://redis.io/blog/benchmarking-results-for-vector-databases/). Vector fields enable you to perform semantic search based on vector embeddings in combination with text, numerical, geospatial, or tag metadata.

**Just looking to get started?** Checkout the vector [quickstart guide]({{< baseurl >}}/develop/get-started/vector-database) and the [Redis AI Resources](https://github.com/redis-developer/redis-ai-resources) repo for more assistance.

## Overview

> 1. [**Create a vector index**]({{< baseurl >}}/develop/interact/search-and-query/advanced-concepts/vectors#create-a-vector-index): Redis maintains a secondary index on top of your data with a defined schema (including vector fields and metadata). Redis supports [`FLAT`]({{< baseurl >}}/develop/interact/search-and-query/advanced-concepts/vectors#flat-index) and [`HNSW`]({{< baseurl >}}/develop/interact/search-and-query/advanced-concepts/vectors#hnsw-index) vector index types.

Reviewer 2: this overview is much more clear!

Reviewer 2: only note: might change "and metadata filters" to "and flexible search over all additional fields in HASH or JSON" 

My thought is this helps highlight the flexibility of those datatypes as a feature instead of a quirk to redis if that makes sense

> 1. [**Store and update vectors**]({{< baseurl >}}/develop/interact/search-and-query/advanced-concepts/vectors#store-and-update-vectors): Redis stores vectors in objects as either Hash or JSON.

Reviewer 3: "Redis stores vectors in objects as either Hash or JSON."
Is it clear from this sentence that these objects may contain more than just vector(s)?

| `field_name`      | Name of the vector field.  |
| `algorithm`       | Vector index algorithm (`FLAT` or `HNSW`).  |
| `index_attribute_count` | Number of vector field attributes.  |
| `index_attribute_name`  | Vector field attribute name.  |
| `index_attribute_value` | Vector field attribute value.  |

### FLAT index
Choose the FLAT index when you have small datasets (< 1M vectors) or when search accuracy is most important, even at the expense of additional latency.

**Required attributes**

| Attribute          | Description                              |
|:-------------------|:-----------------------------------------|
| `TYPE`             | Vector type (`FLOAT16`, `FLOAT32`, `FLOAT64`).  |
| `DIM`              | Vector dimension. Must be a positive integer; identical for both document and query vectors.  |
| `DISTANCE_METRIC`  | Distance metric (`L2`, `IP`, `COSINE`).  |

**Optional attributes**

| Attribute          | Description                                      |
|:-------------------|:-------------------------------------------------|
| `INITIAL_CAP`      | Initial vector capacity in the index.            |

> | `BLOCK_SIZE`       | Block size for memory allocation. Defaults to 1024.  |

Reviewer 3: Consider dropping these from the docs, as we see it as internal tunning that should not have been exposed in the first place (and we plan to remove it)

Author: I will drop them from the documentation!

- At runtime, the search traverses the graph on each layer from top to bottom, finding the local minima before dropping to the subsequent layer.

Choose the HNSW index when you have larger datasets or when search performance and scalability are most important, even at the expense of lower search accuracy.

**Required attributes**

| Attribute          | Description                              |
|:-------------------|:-----------------------------------------|
| `TYPE`             | Vector type (`FLOAT16`, `FLOAT32`, `FLOAT64`). |
| `DIM`              | Vector dimension. *Must be a positive integer.* |
| `DISTANCE_METRIC`  | Distance metric (`L2`, `IP`, `COSINE`).  |

**Optional attributes**

HNSW supports a number of additional parameters to tune
the accuracy of the queries, while trading off performance. Read more about
them [here](https://arxiv.org/ftp/arxiv/papers/1603/1603.09320.pdf).

| Attribute          | Description                                                                                 |
|:-------------------|:--------------------------------------------------------------------------------------------|

> | `INITIAL_CAP`      | Initial vector capacity in the index.                                                           |

Reviewer 3: Same here for the `initial_cap` - consider dropping it

    EPSILON 0.8
```

### Distance metrics

Redis supports three popular distance metrics to measure the degree of similarity between two vectors $u$, $v$ $\in \mathbb{R}^n$ where $n$ is the length of the vectors:

| Distance metric | Description | Mathematical representation |
|:--------------- |:----------- |:--------------------------- |
| `L2` | Euclidean distance between two vectors. | $d(u, v) = \sqrt{ \displaystyle\sum_{i=1}^n{(u_i - v_i)^2}}$ |
| `IP` | Inner product of two vectors. | $d(u, v) = 1 -u\cdot v$ |
| `COSINE` | Cosine distance of two vectors. | $d(u, v) = 1 -\frac{u \cdot v}{\lVert u \rVert \lVert v  \rVert}$ |

The above metrics calculate distance between two vectors, where the smaller the value is, the closer the two vectors are in vector space.

## Store and update vectors

On index creation, the `<storage_type>` dictates how vector and metadata are structured and loaded into Redis.

### Hash
Store or update vectors and any metadata in [hashes]({{< baseurl >}}/develop/data-types/hashes/) with the [`HSET`]({{< baseurl >}}/commands/hset/) command.

**Example**
```
HSET docs:01 vector <vector_bytes> foo bar
```

{{% alert title="Tip" color="warning" %}}

> Hashes expect all data to be represented as strings. Thus, `<vector_bytes>` represents a bytes version of the vector.

Reviewer 4: Consider adding that those strings are binary safe. Consider something like

`Hash values are stored as binary safe strings. Thus, `<vector_bytes>` represents the underlying memory buffer of the vector.`

{{% /alert  %}}

A common way of converting vectors to bytes is through the Python [numpy](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.tobytes.html) library and [redis-py](https://redis-py.readthedocs.io/en/stable/examples/search_vector_similarity_examples.html) client:

```py
import numpy as np
from redis import Redis

redis_client = Redis(host='localhost', port=6379)

# Create a FLOAT32 vector
vector = np.array([0.34, 0.63, -0.54, -0.69, 0.98, 0.61], dtype=np.float32)

# Convert vector to bytes
vector_bytes = vector.tobytes()

# Use redis client to store the vector bytes and metadata under a specified key
redis_client.hset('docs:01', mapping = {"vector": vector_bytes, "foo": "bar"})
```

{{% alert title="Tip" color="warning" %}}

> Note that the vector blob size must match the vector field dimension and float type specified in the schema, otherwise the indexing will fail in the background.

Reviewer 4: `Note that the vector blob size must match the vector field dimension and float type specified in the schema, otherwise the indexing will fail`

{{% /alert  %}}

### JSON
Store or update vectors and any metadata in [JSON]({{< baseurl >}}/develop/data-types/json/) with the [`JSON.SET`]({{< baseurl >}}/commands/json.set/) command.

Unlike in hashes, vectors are stored in JSON documents as arrays (not as bytes).

**Example**
```
JSON.SET docs:01 $ '{"vector":[0.34,0.63,-0.54,-0.69,0.98,0.61], "foo": "bar"}'
```

One of the additional benefits of JSON is schema flexibility. As of v2.6.1, JSON supports multi-value indexing.
Thus, it is possible to index multiple vectors under the same JSONPath.

Here are some examples of multi-value indexing with vectors:

**Example**
```
JSON.SET key $ '{"vector":[[1,2,3,4], [5,6,7,8]]}'
JSON.SET key $ '{"item1":{"vector":[1,2,3,4]}, "item2":{"vector":[5,6,7,8]}}'
```

Additional information and examples are available in the [Indexing JSON documents]({{< baseurl >}}/develop/interact/search-and-query/indexing/#index-json-arrays-as-vector) section.

## Search with vectors
You can run vector search queries with the [`FT.SEARCH`]({{< baseurl >}}/commands/ft.search/) or [`FT.AGGREGATE`]({{< baseurl >}}/commands/ft.aggregate/) commands.

> To use a vector similarity query with `FT.SEARCH`, you must always specify the `DIALECT 2` option. See the [dialects page]({{< relref "/develop/interact/search-and-query/advanced-concepts/dialects" >}}) for more information.

Reviewer 4: can be used with all dialects >= 2

### Standard vector search

Standard vector search has the following common syntax:

**Syntax**

```
FT.SEARCH <index_name>
  <primary_filter_query>=>[KNN <top_k> @<vector_field> $<vector_blob_param> $<vector_query_params> AS <distance_field>]
  PARAMS <vector_query_params_count> [<vector_query_param_name> <vector_query_param_value> ...]
  SORTBY <distance_field>
  DIALECT 2
```

> TODO ^^ Need to validate this query generalization

Reviewer 4: @tylerhutcherson `PARAMS` is a general purpose keyword for parameterized queries so `<vector_query_params_{word}>` need to be changed to `<query_params_{word}>` . Also you must send your blob here (not present in the above example)

**Parameters**

| Parameter         | Description                                                                                       |
|:------------------|:--------------------------------------------------------------------------------------------------|
| `index_name`  | Name of the index.  |

> | `primary_filter_query`  | Optional [filter]({{< baseurl >}}/develop/interact/search-and-query/advanced-concepts/vectors#filters) criteria. Defaults to `*`.  |

Reviewer 3: Should we say "defaults to *"? It may imply that you don't have to explicitly write * for non-filtering semantics (which you do...)

Reviewer 4: I think it should not be marked as optional

> | `top_k` | Number of nearest neighbors. Also commonly provided to the `LIMIT` param to truncate the number of results through pagination.  |

Reviewer 4: what do you think about specifying that `LIMIT` is "stronger" than `top_k` since it is global (like `top_k` = 1000 and `LIMIT`=20, will yield 20 results).

| `vector_blob_param`  | The query vector as bytes and must be passed through the `PARAMS` section. The blob's byte size should match the vector field dimension and type.  |
| `vector_query_params` (optional) | An optional section for marking one or more vector query parameters passed through the `PARAMS` section. Valid parameters should be provided as key-value pairs. See which [runtime query params]({{< baseurl >}}/develop/interact/search-and-query/advanced-concepts/vectors#runtime-query-params) are supported for each vector index type.  |
| `distance_field` (optional) | The optional distance field name used in the response and/or for sorting. By default, the distance field name is "`__<vector_field>_score`" and it can be used for sorting without using `AS <distance_field>` in the query.  |
| `vector_query_params_count` | The number of vector query params.  |
| `vector_query_param_name` | The name of the vector query param.  |
| `vector_query_param_value` | The value of the vector query param.  |

**Example**

```
FT.SEARCH my_index "(@title:Matrix @year:[2020 2022])=>[KNN 10 @vector $BLOB]" PARAMS 2 BLOB "\x12\xa9\xf5\x6c" DIALECT 2
```

**Use query attributes**

> Alternatively, as of v2.6, `<vector_query_params> and `<distance_field>` name can be specified in runtime

Reviewer 3: Missing "`" before "and"

**Supported filter types**

- [Exact Match Filters](https://redis.io/docs/develop/interact/search-and-query/query/exact-match/)
- [Numeric Range Filters](https://redis.io/docs/develop/interact/search-and-query/query/range/)
- [Full-text Search](https://redis.io/docs/develop/interact/search-and-query/query/full-text/)
- [Geospatial](https://redis.io/docs/develop/interact/search-and-query/query/geo-spatial/)

You can also [combine multiple queries](https://redis.io/docs/develop/interact/search-and-query/query/combined/) as a filter.

**Syntax**

Vector search queries with filters follow this basic structure:

```
FT.SEARCH <index_name> <primary_filter_query>=>[...]
```

where `<primary_filter_query>` defines document selection and filtering.

### How Filtering Works

Redis uses internal algorithms to optimize the filtering computation for vector search. The runtime algorithm is determined by heuristics that aim to minimize query latency, based on several factors derived from the query and the index.

**Batches mode**

> - A batch of high-scoring documents from the vector index is retrieved. These documents are yielded ONLY if the `<primary_filter_query>` is satisfied. In other words, the document must contain a similar vector and meet the filter criteria.

Reviewer 4: Consider comparing it to paginating over the vector index. I think it will give people the right context.

- oss
- kubernetes
- clients
description: Learn how to use vector fields and perform vector search in Redis
linkTitle: Vectors
math: true
title: Vectors
weight: 14
---

Redis is commonly used as a [highly-performant vector database](https://redis.io/blog/benchmarking-results-for-vector-databases/). Vector fields enable you to perform semantic search based on vector embeddings in combination with metadata like text, numerics, geo, or tags.

> **Just looking to get started?** Checkout the vector [quickstart guide]({{< baseurl >}}/develop/get-started/vector-database) and the [Redis AI Resources](https://github.com/redis-developer/redis-ai-resources) repo for more assistance.

Author: It is a single repo `https://github.com/redis-developer/redis-ai-resources`

## Overview

> ### Storage options

Reviewer 2: It goes from a very brief overview to realtime indexing without a lot of context here. 

I think adding a brief sentence or two outlining how the pieces work together to enable vector search could be really valuable to level-set.

- redis stores vectors as a field in an object of type hash or json
- to query for vectors you need to create an index for the object and for the vector
- both these indexes come with certain configurations

Generally I think we need something here to show how the piece come together in the end a bit more if that makes sense

Author: Agreed. Updated this @rbs333

When using Redis as a vector database, you have the choice between storing vectors and metadata in a Hash or JSON object.

| Storage option             | Description                          |
|-------------------------|--------------------------------------|

> | **Hash** | Lightweight, single-level dictionary of fields. Most efficient choise for memory optimization and query speed.    |

Reviewer 5: ```suggestion
| **Hash** | Lightweight, single-level dictionary of fields. Most efficient choice for memory optimization and query speed.    |
```

**Syntax**

```
FT.CREATE <index_name>
  ON <storage_type>
  PREFIX 1 <key_prefix>
  SCHEMA ... <field_name> VECTOR <algorithm> <index_attribute_count>
    [<index_attribute_name> <index_attribute_value> ...]
```

Refer to the full [indexing]({{< baseurl >}}/develop/interact/search-and-query/indexing/) documentation for additional fields, option, and noted limitations.

**Parameters**

| Parameter         | Description                                                                                       |
|-------------------|---------------------------------------------------------------------------------------------------|
| `index_name`      | Name of the index.  |
| `storage_type`    | Storage option (`HASH` or `JSON`).  |
| `prefix` (optional) | Key prefix that signals to the index which keys should be included. Defaults to all keys if omitted. |
| `field_name`      | Name of the vector field.  |
| `algorithm`       | Vector index algorithm (`FLAT` or `HNSW`).  |
| `index_attribute_count` | Number of vector field attributes.  |
| `index_attribute_name`  | Vector field attribute name.  |
| `index_attribute_value` | Vector field attribute value.  |

### FLAT index

> **Choose the FLAT index when** you have small datasets (*<1M vectors*) or when search accuracy is most important *even at the expense of additional latency*.

Reviewer 2: I like the tips you add in I think that's an improvement and helps orient the reader

| `DIM`              | Vector dimension. *Must be a positive integer; identical for both document and query vectors.*  |
| `DISTANCE_METRIC`  | Distance metric (`L2`, `IP`, `COSINE`).  |

**Optional attributes**

| Attribute          | Description                                      |
|--------------------|--------------------------------------------------|
| `INITIAL_CAP`      | Initial vector capacity in the index.            |
| `BLOCK_SIZE`       | Block size for memory allocation. Defaults to 1024.  |

**Example**

```
FT.CREATE my_index
  ON HASH
  PREFIX 1 docs:
  SCHEMA my_vector_field VECTOR FLAT 6
    TYPE FLOAT32
    DIM 1536
    DISTANCE_METRIC COSINE
```
In the example above, we create an index named `my_index` over hashes with the key prefix `docs:`, including a `FLAT` vector field named `my_vector_field` with 3 index attributes (`TYPE`, `DIM`, and `DISTANCE_METRIC`).

### HNSW index

HNSW is an approximate nearest neighbors algorithm (ANN) that uses a multi-layered graph to make vector search more scalable.
- The lowest layer contains all data points, and each higher layer contains a subset, forming a hierarchy.

> - At runtime, the search greedily traverses the graph on each layer from top to bottom, finding the "local minima" before dropping to the subsequent layer.

Reviewer 5: ```suggestion
- At runtime, the search traverses the graph on each layer from top to bottom, finding the "local minima" before dropping to the subsequent layer.
```

**Choose the HNSW index when** you have larger datasets or when search performance and scalability are most important *even at the expense of lower search accuracy*.

**Required attributes**

| Attribute          | Description                              |
|--------------------|------------------------------------------|
| `TYPE`             | Vector type (`FLOAT16`, `FLOAT32`, `FLOAT64`). |
| `DIM`              | Vector dimension. *Must be a positive integer.* |
| `DISTANCE_METRIC`  | Distance metric (`L2`, `IP`, `COSINE`).  |

**Optional attributes**

> HNSW supports a number of additional params to tune

Reviewer 5: ```suggestion
HNSW supports a number of additional parameters to tune
```

the accuracy of the queries, while trading off performance. Read more about
them [here](https://arxiv.org/ftp/arxiv/papers/1603/1603.09320.pdf).

| Attribute          | Description                                                                                     |
|--------------------|-------------------------------------------------------------------------------------------------|
| `INITIAL_CAP`      | Initial vector capacity in the index.                                                           |
| `M`                | Max number of outgoing edges (connections) for each node in the graph per layer. On layer zero the max number of connections will be 2M. Higher values increase accuracy, but also increase memory usage and index build time. Defaults to 16.            |
| `EF_CONSTRUCTION`  | Max number of connected neighbors to consider during graph building. Higher values increase accuracy, but also increase index build time. Defaults to 200.                                      |
| `EF_RUNTIME`       | Max top candidates during KNN search. Higher values increase accuracy, but also increase search latency. Defaults to 10. |

> | `EPSILON`          | Relative factor that sets the boundaries in which a range query may search for candidates. That is, vector candidates whose distance from the query vector is radius*(1 + EPSILON) are potentially scanned, allowing more extensive search and more accurate results (on the expense of runtime). Defaults to 0.01.                                             |

Reviewer 5: ```suggestion
| `EPSILON`          | Relative factor that sets the boundaries in which a range query may search for candidates. That is, vector candidates whose distance from the query vector is radius*(1 + EPSILON) are potentially scanned, allowing more extensive search and more accurate results (at the expense of runtime). Defaults to 0.01.                                             |
```

| `IP` | Inner product of two vectors. | $d(u, v) = 1 -u\cdot v$ |
| `COSINE` | Cosine distance of two vectors. | $d(u, v) = 1 -\frac{u \cdot v}{\lVert u \rVert \lVert v  \rVert}$ |

*The above metrics calculate distance between two vectors, where the smaller the value is, the closer the two vectors are in vector space.*

## Store and update vectors

On index creation, the `<storage_type>` dictates how vector and metadata are structured and loaded into Redis.

### Hash
Store or update vectors and any metadata in [hashes]({{< baseurl >}}/develop/data-types/hashes/) with the [`HSET`]({{< baseurl >}}/commands/hset/) command.

**Example**
```
HSET docs:01 vector <vector_bytes> foo bar
```

<br/>

{{% alert title="Tip" color="warning" %}}
Hashes expect all data to be represented as strings. Thus, `<vector_bytes>` represents a bytes version of the vector.
{{% /alert  %}}

<br/>

A common way of converting vectors to bytes is through the Python [numpy](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.tobytes.html) library and [redis-py](https://redis-py.readthedocs.io/en/stable/examples/search_vector_similarity_examples.html) client:
```py
import numpy as np

> from redis import Redis

Reviewer 2: something is a bit odd to me about jumping to a python example if all other example are just cli. I think the long term plan is to have each example be a carousel of client libary examples but if that's not till the future may consider making this just a note or link to something else

Author: Yeah, agreed. This was in place before, I just tried to improve the Python ;)

redis_client = Redis(host='localhost', port=6379)

# Create a FLOAT32 vector
vector = np.array([0.34, 0.63, -0.54, -0.69, 0.98, 0.61], dtype=np.float32)

# Convert vector to bytes
vector_bytes = vector.tobytes()

# Use redis client to store the vector bytes and metadata under a specified key
redis_client.hset('docs:01', mapping = {"vector": vector_bytes, "foo": "bar"})
```
<br/>
{{% alert title="Tip" color="warning" %}}
Note that the vector blob size must match the vector field dimension and float type specified in the schema, otherwise the indexing will fail in the background.
{{% /alert  %}}
<br/>

### JSON
Store or update vectors and any metadata in [JSON]({{< baseurl >}}/develop/data-types/json/) with the [`JSON.SET`]({{< baseurl >}}/commands/json.set/) command.

Unlike in hashes, vectors are stored in JSON documents as arrays (not as bytes).

**Example**
```
JSON.SET docs:01 $ '{"vector":[0.34,0.63,-0.54,-0.69,0.98,0.61], "foo": "bar"}'
```

**One of the additional benefits of JSON is schema flexibility.** As of v2.6.1, JSON supports multi-value indexing.
Thus, it is possible to index multiple vectors under the same JSONPath.

> Here are a some examples of multi-value indexing with vectors:

Reviewer 2: drop a

**Example**
```
JSON.SET key $ '{"vector":[[1,2,3,4], [5,6,7,8]]}'
JSON.SET key $ '{"item1":{"vector":[1,2,3,4]}, "item2":{"vector":[5,6,7,8]}}'
```

Additional information and examples are available in the [Indexing JSON documents]({{< baseurl >}}/develop/interact/search-and-query/indexing/#index-json-arrays-as-vector) section.

## Search with vectors
You can run vector search queries with the [`FT.SEARCH`]({{< baseurl >}}/commands/ft.search/) or [`FT.AGGREGATE`]({{< baseurl >}}/commands/ft.aggregate/) commands.

To use a vector similarity query with `FT.SEARCH`, you must always:
- Specify the option `DIALECT 2` or greater in the search command at runtime.

> - Or, set the `DEFAULT_DIALECT` option to `2` or greater by modifying the search configuration with [`FT.CONFIG SET`]({{< baseurl >}}/commands/ft.config-set/).

Reviewer 2: I think this can be made slightly more prominant with an explination of why dialects exist. I had a customer looking at one of our docs sets and was super confused by the dialect things so that's where this ask is coming from

Author: Yeah, we should link out to the general section on dialects, because it's definitely a foot gun

| `vector_blob_param`  | The query vector as bytes and must be passed through the `PARAMS` section. The blob's byte size should match the vector field dimension and type.  |
| `vector_query_params` (optional) | An optional section for marking one or more vector query parameters passed through the `PARAMS` section. Valid parameters should be provided as key-value pairs. See which [runtime query params]({{< baseurl >}}/develop/interact/search-and-query/advanced-concepts/vectors#runtime-query-params) are supported for each vector index type.  |
| `distance_field` (optional) | The optional distance field name used in the response and/or for sorting. By default, the distance field name is "`__<vector_field>_score`" and it can be used for sorting without using `AS <distance_field>` in the query.  |
| `vector_query_params_count` | The number of vector query params.  |
| `vector_query_param_name` | The name of the vector query param.  |
| `vector_query_param_value` | The value of the vector query param.  |

**Example**

```
FT.SEARCH my_index "(@title:Matrix @year:[2020 2022])=>[KNN 10 @my_vector_field $BLOB]" PARAMS 2 BLOB "\x12\xa9\xf5\x6c" DIALECT 2
```

> **Invalid example**

Reviewer 2: is this a super common mistake? I don't love providing invalid examples because as an engineer I see that as: "our api is confusing so people intuitievly use it wrong... sorry"

Author: great point

Misplaced filter statement, missing `BLOB` param, and missing dialect.
```
FT.SEARCH my_index "(@title:Matrix)=>[KNN 10 @my_vector_field $BLOB] @year:[2020 2022]"
```

**Using query attributes**

> Alternatively, as of v2.6, `<vector_query_params> and `<distance_field>` name can be specified in runtime [query attributes]({<< baseurl >}}/develop/interact/search-and-query/advanced-concepts/query_syntax#query-attributes) as well. Thus, the following format is also supported:

Reviewer 5: ```suggestion
Alternatively, as of v2.6, `<vector_query_params> and `<distance_field>` name can be specified in runtime [query attributes]({<< baseurl >}}/develop/interact/search-and-query/advanced-concepts/query_syntax/#query-attributes) as well. Thus, the following format is also supported:
```

| `vector_query_params` (optional) | An optional section for marking one or more vector query parameters passed through the `PARAMS` section. Valid parameters should be provided as key-value pairs. See which [runtime query params]({{< baseurl >}}/develop/interact/search-and-query/advanced-concepts/vectors#runtime-query-params) are supported for each vector index type.  |
| `vector_query_params_count` | The number of vector query params.  |
| `vector_query_param_name` | The name of the vector query param.  |
| `vector_query_param_value` | The value of the vector query param.  |

**Using query attributes**

Vector range queries clause can be followed by a query attributes section as follows: `@<vector_field>: [VECTOR_RANGE (<radius> | $<radius_param>) $<vector_blob_param>]=>{$<param>: (<value> | $<value_attribute>); ... }`, where the relevant params in that case are `$yield_distance_as` and `$epsilon`. Note that there is **no default distance field name** in range queries.

TODO ^^ need to clean this up a bit more and tighten the language.

Sure, here’s the revised and improved version of the section on filtering and supported modes:

---

### Filters

> Redis supports vector search along with various filters to narrow the search space based on defined criteria. If your index contains searchable fields (i.e., `TEXT`, `TAG`, `NUMERIC`, `GEO`, `GEOSHAPE`, and `VECTOR`), you can perform vector searches with filters.

Reviewer 5: ```suggestion
Redis supports vector search along with various filters to narrow the search space based on defined criteria. If your index contains searchable fields (for example, `TEXT`, `TAG`, `NUMERIC`, `GEO`, `GEOSHAPE`, and `VECTOR`), you can perform vector searches with filters.
```

```
Use [query attributes]({{< baseurl >}}/develop/interact/search-and-query/advanced-concepts/query_syntax#query-attributes) syntax to specify optional parameters and the distance field name:
```
FT.SEARCH idx "*=>[KNN 10 @vector $BLOB]=>{$EF_RUNTIME: $EF; $YIELD_DISTANCE_AS: my_scores}" PARAMS 4 EF 150 BLOB "\x12\xa9\xf5\x6c" SORTBY my_scores DIALECT 2
```

Find additional Python examples for [`redis-py`](https://github.com/redis-developer/redis-ai-resources/blob/main/python-recipes/vector-search/00_redispy.ipynb) and [`redisvl`](https://github.com/redis-developer/redis-ai-resources/blob/main/python-recipes/vector-search/01_redisvl.ipynb).

### Pre-filter KNN queries (hybrid approach)

Among documents that have `'Dune'` in their `title` field and their `num` value is in the range `[2020, 2022]`, return the top 10 for which the vector stored in the `vector` field is the closest to the vector represented by the following 4-bytes blob:

```
FT.SEARCH idx "(@title:Dune @num:[2020 2022])=>[KNN $K @vector $BLOB AS my_scores]" PARAMS 4 BLOB "\x12\xa9\xf5\x6c" K 10 SORTBY my_scores DIALECT 2
```

> Use a different filter for the hybrid approach: this time, return the top 10 results from the documents that contain a `'shirt'` tag  in the `type` field and optionally a `'blue'` tag in their `color` field. Here, the results are sorted by the full-text scorer.

Reviewer 5: ```suggestion
Use a different filter for the hybrid approach: this example returns the top 10 results from the documents that contain a `'shirt'` tag  in the `type` field and optionally a `'blue'` tag in their `color` field. The results are sorted by the full-text scorer.
```

```
FT.SEARCH idx "(@type:{shirt} ~@color:{blue})=>[KNN $K @vector $BLOB]" PARAMS 4 BLOB "\x12\xa9\xf5\x6c" K 10 DIALECT 2
```

> And, here's a pre-filter with KNN query in which the hybrid policy is set explicitly to "ad-hoc brute force" (rather than auto-selected):

Reviewer 5: ```suggestion
This example shows a pre-filter with KNN query in which the hybrid policy is set explicitly to "ad-hoc brute force" (rather than auto-selected):
```

```
FT.SEARCH idx "(@type:{shirt})=>[KNN $K @vector $BLOB HYBRID_POLICY ADHOC_BF]" PARAMS 4 BLOB "\x12\xa9\xf5\x6c" K 10 SORTBY __vec_scores DIALECT 2
```

> And, now, here's a pre-filter with KNN query in which the hybrid policy is set explicitly to "batches", and the batch size is set explicitly to be 50 using a query parameter:

Reviewer 5: ```suggestion
This example shows a pre-filter with KNN query in which the hybrid policy is set explicitly to "batches", and the batch size is set explicitly to be 50 using a query parameter:
```

weight: 14
---

Redis includes a [high performance vector database](https://redis.io/blog/benchmarking-results-for-vector-databases/) that lets you perform semantic search over vector embeddings. You can augment these searches with filtering over text, numerical, geospatial, and tag metadata.

> To quickly get started, check out out the [Redis vector quickstart guide]({{< baseurl >}}/develop/get-started/vector-database) and the [Redis AI Resources](https://github.com/redis-developer/redis-ai-resources) Github repo.

Reviewer: ```suggestion
To quickly get started, check out the [Redis vector quickstart guide]({{< baseurl >}}/develop/get-started/vector-database) and the [Redis AI Resources](https://github.com/redis-developer/redis-ai-resources) Github repo.
```
